### ISSUE TITLE
Bug: Mysql Appliance has a really big delay and performance variance

### ISSUE THREAD

**Issue by onnimonni** [flynn/flynn#3291]
**TL;DR:** Flynn Mysql Appliance is ~15 times slower than local default mariadb container.

**EDIT:** `my.cnf` configs have nothing to do with this. This performance variance doesn't exist for [default mariadb docker container](https://hub.docker.com/_/mariadb/).

I tried running WordPress in docker container with the default mysql appliance and it is super slow even though I'm using redis as object cache.

Here's some metrics:

**Connection from container inside Flynn to Mariadb Appliance:**

``` sh
onnimonni@local: $ flynn -a my-wordpress-app run bash
bash-4.3# mysqlslap -h $MYSQL_HOST --password=$MYSQL_PWD -u $MYSQL_USER
mysqlslap: Error when connecting to server: 1044 Access denied for user 'c595cd303ee2d2b69c12030f5bd7167d'@'%' to database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 15.035 seconds
    Minimum number of seconds to run all queries: 15.035 seconds
    Maximum number of seconds to run all queries: 15.035 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0
bash-4.3# mysqlslap -h $MYSQL_HOST --password=$MYSQL_PWD -u $MYSQL_USER
mysqlslap: Error when connecting to server: 1044 Access denied for user 'c595cd303ee2d2b69c12030f5bd7167d'@'%' to database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 7.018 seconds
    Minimum number of seconds to run all queries: 7.018 seconds
    Maximum number of seconds to run all queries: 7.018 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h $MYSQL_HOST --password=$MYSQL_PWD -u $MYSQL_USER
mysqlslap: Error when connecting to server: 1044 Access denied for user 'c595cd303ee2d2b69c12030f5bd7167d'@'%' to database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.012 seconds
    Minimum number of seconds to run all queries: 1.012 seconds
    Maximum number of seconds to run all queries: 1.012 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h $MYSQL_HOST --password=$MYSQL_PWD -u $MYSQL_USER
mysqlslap: Error when connecting to server: 1044 Access denied for user 'c595cd303ee2d2b69c12030f5bd7167d'@'%' to database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 2.011 seconds
    Minimum number of seconds to run all queries: 2.011 seconds
    Maximum number of seconds to run all queries: 2.011 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0
```

**EDIT:**
**I tested also the connection inside mysql container to see if this is zfs problem:**

```
onnimonni@local: $ flynn -a mariadb run bash
root@0ec1b2b0-b2d6-436b-ac59-35bed3826fe5:/tmp# mysqlslap -h $MYSQL_HOST --password=$MYSQL_PWD -u $MYSQL_USER
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 4.017 seconds
    Minimum number of seconds to run all queries: 4.017 seconds
    Maximum number of seconds to run all queries: 4.017 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

root@0ec1b2b0-b2d6-436b-ac59-35bed3826fe5:/tmp# mysqlslap -h $MYSQL_HOST --password=$MYSQL_PWD -u $MYSQL_USER
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.013 seconds
    Minimum number of seconds to run all queries: 1.013 seconds
    Maximum number of seconds to run all queries: 1.013 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

```

These don't include the waiting time `mysqlslap` needs to have before it can start the tests.

Maybe this is related to the overlay network?

**I ran `mysqltuner.pl` inside Flynn wordpress container to inspect the needed changes to achieve better performance**
**Here's the output:**

```
bash-4.3# ./mysqltuner.pl --host $MYSQL_HOST --user $MYSQL_USER --pass $MYSQL_PWD --forcemem 2048
 >>  MySQLTuner 1.6.15 - Major Hayden <major@mhtx.net>
 >>  Bug reports, feature requests, and downloads at http://mysqltuner.com/
 >>  Run with '--help' for additional options and output filtering

[--] Skipped version check for MySQLTuner script
[--] Performing tests on leader.mariadb.discoverd:3306
[OK] Logged in using credentials passed on the command line
[--] Assuming 2048 MB of physical memory
[!!] Assuming 0 MB of swap space (use --forceswap to specify)
[!!] failed to execute: SHOW SLAVE STATUS\G
[!!] FAIL Execute SQL / return code: 256
[!!] failed to execute: SHOW SLAVE HOSTS
[!!] FAIL Execute SQL / return code: 256
[OK] Currently running supported MySQL version 10.1.16-MariaDB-1~trusty

-------- Storage Engine Statistics -----------------------------------------------------------------
[--] Status: +Aria +CSV +InnoDB +MEMORY +MRG_MyISAM +MyISAM +PERFORMANCE_SCHEMA +SEQUENCE
[--] Data in InnoDB tables: 848K (Tables: 20)
[OK] Total fragmented tables: 0

-------- Security Recommendations ------------------------------------------------------------------
[!!] failed to execute: SELECT CONCAT(user, '@', host) FROM mysql.user WHERE TRIM(USER) = '' OR USER IS NULL
[!!] FAIL Execute SQL / return code: 256
[OK] There are no anonymous accounts for any database users
[!!] failed to execute: SELECT CONCAT(user, '@', host) FROM mysql.user WHERE (password = '' OR password IS NULL) AND plugin NOT IN ('unix_socket', 'win_socket')
[!!] FAIL Execute SQL / return code: 256
[OK] All database users have passwords assigned
[!!] failed to execute: SELECT CONCAT(user, '@', host) FROM mysql.user WHERE CAST(password as Binary) = PASSWORD(user) OR CAST(password as Binary) = PASSWORD(UPPER(user)) OR CAST(password as Binary) = PASSWORD(UPPER(LEFT(User, 1)) + SUBSTRING(User, 2, LENGTH(User)))
[!!] FAIL Execute SQL / return code: 256
[!!] failed to execute: SELECT CONCAT(user, '@', host) FROM mysql.user WHERE HOST='%'
[!!] FAIL Execute SQL / return code: 256
[!!] There is no basic password file list!

-------- CVE Security Recommendations --------------------------------------------------------------
[--] Skipped due to --cvefile option undefined

-------- Performance Metrics -----------------------------------------------------------------------
[--] Up for: 2d 21h 33m 51s (17K q [0.070 qps], 4K conn, TX: 2M, RX: 1M)
[--] Reads / Writes: 91% / 9%
[--] Binary logging is enabled (GTID MODE: OFF)
[--] Physical Memory     : 2.0G
[--] Max MySQL memory    : 1.3G
[--] Other process memory: 0B
[--] Total buffers: 624.0M global + 7.5M per thread (100 max threads)
[--] P_S Max memory usage: 0B
[--] Galera GCache Max memory usage: 0B
[OK] Maximum reached memory usage: 654.1M (31.94% of installed RAM)
[OK] Maximum possible memory usage: 1.3G (67.25% of installed RAM)
[OK] Overall possible memory usage with other process is compatible with memory available
[OK] Slow queries: 0% (0/17K)
[OK] Highest usage of available connections: 4% (4/100)
[OK] Aborted connections: 0.20%  (8/4049)
[!!] name resolution is active : a reverse name resolution is made for each new connection and can reduce performance
[!!] Query cache efficiency: 6.1% (678 cached / 11K selects)
[OK] Query cache prunes per day: 0
[OK] Sorts requiring temporary tables: 0% (0 temp sorts / 63 sorts)
[OK] No joins without indexes
[!!] Temporary tables created on disk: 97% (1K on disk / 1K total)
[OK] Table cache hit rate: 55% (29 open / 52 opened)
[OK] Open file limit used: 0% (25/10K)
[OK] Table locks acquired immediately: 100% (3K immediate / 3K locks)
[OK] Binlog cache memory access: 100.00% ( 983 Memory / 983 Total)

-------- Performance schema ------------------------------------------------------------------------
[--] Performance schema is disabled.

-------- ThreadPool Metrics ------------------------------------------------------------------------
[--] ThreadPool stat is enabled.
[--] Thread Pool Size: 2 thread(s).
[--] Using default value is good enough for your version (10.1.16-MariaDB-1~trusty)

-------- MyISAM Metrics ----------------------------------------------------------------------------
[!!] Key buffer used: 18.2% (24M used / 134M cache)
[!!] Cannot calculate MyISAM index size - re-run script as root user

-------- AriaDB Metrics ----------------------------------------------------------------------------
[--] AriaDB is enabled.
[OK] Aria pagecache size / total Aria indexes: 128.0M/1B
[!!] Aria pagecache hit rate: 85.7% (1K cached / 177 reads)

-------- InnoDB Metrics ----------------------------------------------------------------------------
[--] InnoDB is enabled.
[OK] InnoDB buffer pool / data size: 256.0M/848.0K
[!!] InnoDB buffer pool <= 1G and innodb_buffer_pool_instances(!=1).
[!!] InnoDB Used buffer: 3.24% (531 used/ 16383 total)
[OK] InnoDB Read buffer efficiency: 99.55% (35973 hits/ 36134 total)
[!!] InnoDB Write Log efficiency: 67.94% (2666 hits/ 3924 total)
[OK] InnoDB log waits: 0.00% (0 waits / 1258 writes)

-------- TokuDB Metrics ----------------------------------------------------------------------------
[--] TokuDB is disabled.

-------- Galera Metrics ----------------------------------------------------------------------------
[--] Galera is disabled.

-------- Replication Metrics -----------------------------------------------------------------------
[--] Galera Synchronous replication: NO
[--] No replication slave(s) for this server.
[--] This is a standalone server.

-------- Recommendations ---------------------------------------------------------------------------
General recommendations:
    Configure your accounts with ip or subnets only, then update your configuration with skip-name-resolve=1
    When making adjustments, make tmp_table_size/max_heap_table_size equal
    Reduce your SELECT DISTINCT queries which have no LIMIT clause
Variables to adjust:
    query_cache_limit (> 128K, or use smaller result sets)
    tmp_table_size (> 32M)
    max_heap_table_size (> 32M)
    innodb_buffer_pool_instances (=1)
```

I would use at least:

```
skip-name-resolve=1
tmp_table_size=64M
max_heap_table_size=64M
```

But the most important rule would be increasing `query_cache_limit=128M`. At least for my setup it should be as high as possible.

Latency test between nodes is quite good:

```
$ ping node2
PING node2 (node2) 56(84) bytes of data.
64 bytes from node2: icmp_seq=1 ttl=64 time=0.178 ms
64 bytes from node2: icmp_seq=2 ttl=64 time=0.143 ms
64 bytes from node2: icmp_seq=3 ttl=64 time=0.145 ms
64 bytes from node2: icmp_seq=4 ttl=64 time=0.153 ms
64 bytes from node2: icmp_seq=5 ttl=64 time=0.152 ms
64 bytes from node2: icmp_seq=6 ttl=64 time=0.143 ms
64 bytes from node2: icmp_seq=7 ttl=64 time=0.152 ms
```

My flynn nodes have 2gb of memory, the minimum recommended amount. Is this not enough?

**Comment by onnimonni** [comment]
In addition I'm using default zfs options from installation guide + these tweaks:
```
atime=off
ashift=12
```

**Comment by onnimonni** [comment]
I thought that this is related to the mysql configs but it's somehow related how the appliances are networked.

I tried to run vanilla mariadb docker container in flynn and it was just fine:
```bash
# First create custom docker database
$ flynn create my-custom-database --remote=""
$ flynn -a my-custom-database docker push mariadb

# Then log to my other container and try the tests again
$ flynn -a wordpress run bash
$ mysqlslap -h my-custom-database-web.discoverd --user=root --password=root
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.013 seconds
    Minimum number of seconds to run all queries: 1.013 seconds
    Maximum number of seconds to run all queries: 1.013 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h my-custom-database-web.discoverd --user=root --password=root
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.013 seconds
    Minimum number of seconds to run all queries: 1.013 seconds
    Maximum number of seconds to run all queries: 1.013 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h my-custom-database-web.discoverd --user=root --password=root
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.013 seconds
    Minimum number of seconds to run all queries: 1.013 seconds
    Maximum number of seconds to run all queries: 1.013 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h my-custom-database-web.discoverd --user=root --password=root
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.014 seconds
    Minimum number of seconds to run all queries: 1.014 seconds
    Maximum number of seconds to run all queries: 1.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h my-custom-database-web.discoverd --user=root --password=root
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.012 seconds
    Minimum number of seconds to run all queries: 1.012 seconds
    Maximum number of seconds to run all queries: 1.012 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

bash-4.3# mysqlslap -h my-custom-database-web.discoverd --user=root --password=root
mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'
Benchmark
    Average number of seconds to run all queries: 1.013 seconds
    Minimum number of seconds to run all queries: 1.013 seconds
    Maximum number of seconds to run all queries: 1.013 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0
```

I can see that there's no variance at all.

**Comment by onnimonni** [comment]
But here are the results of both `mysqlslap` and `time` for seeing if there's connection delay for mysqldap.
```

# I hope that the results would always look like this

time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 1.014 seconds
    Minimum number of seconds to run all queries: 1.014 seconds
    Maximum number of seconds to run all queries: 1.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m1.020s
user    0m0.000s
sys 0m0.000s
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 1.014 seconds
    Minimum number of seconds to run all queries: 1.014 seconds
    Maximum number of seconds to run all queries: 1.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m1.020s
user    0m0.000s
sys 0m0.000s
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 1.014 seconds
    Minimum number of seconds to run all queries: 1.014 seconds
    Maximum number of seconds to run all queries: 1.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m1.020s
user    0m0.000s
sys 0m0.000s
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 1.014 seconds
    Minimum number of seconds to run all queries: 1.014 seconds
    Maximum number of seconds to run all queries: 1.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m1.019s
user    0m0.001s
sys 0m0.000s
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 2.014 seconds
    Minimum number of seconds to run all queries: 2.014 seconds
    Maximum number of seconds to run all queries: 2.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0


real    0m2.020s
user    0m0.000s
sys 0m0.000s

##
# But sometimes they look like this
##
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 32.255 seconds
    Minimum number of seconds to run all queries: 32.255 seconds
    Maximum number of seconds to run all queries: 32.255 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m32.260s
user    0m0.000s
sys 0m0.000s

##
# This one is super strange
# I think the mysqlslap only tracks the time that it takes after it has connected to the database
##
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 1.013 seconds
    Minimum number of seconds to run all queries: 1.013 seconds
    Maximum number of seconds to run all queries: 1.013 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m16.036s
user    0m0.000s
sys 0m0.000s
bash-4.3# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
    Average number of seconds to run all queries: 1.014 seconds
    Minimum number of seconds to run all queries: 1.014 seconds
    Maximum number of seconds to run all queries: 1.014 seconds
    Number of clients running queries: 1
    Average number of queries per client: 0

real    0m4.021s
user    0m0.000s
sys 0m0.002s
```

**Comment by onnimonni** [comment]
Do you know if the appliances have a different kind of system for networking than applications have?

**Comment by josephglanville** [comment]
If the tests are writing a bunch of data you might be running into #3197. Apart from that I am not sure how to interpret the data you have, I am not familiar with these tools unfortunately.

**Comment by onnimonni** [comment]
So for all of your setups so far mysql appliance has been working fast enough?

These are pretty much the steps to reproduce my problem. Can you see what happens if you do this:

```
# Use mariadb container since it provides mysqlslap tool for testing
$ docker pull mariadb

$ flynn docker login
$ flynn create database-test --remote=""
$ flynn -a database-test docker push mariadb
$ flynn -a database-test resource add mysql
$ flynn -a database-test run bash

# Test the mysql appliance performance few times and see if there's plenty of variance
root@1de11d05f2af:/# time mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD
```

**Comment by josephglanville** [comment]
I tested this benchmark tool out, I don't think you currently have it configured correctly. The reason you constantly see times in excess of 1 second is it's not successfully creating a connection to the correct database, hence the `mysqlslap: Error when connecting to server: 1049 Unknown database 'mysqlslap'`. In the case of Flynn you are likely not going to be able to use the `--auto-generate-schema` with `--create-schema=$MYSQL_DATABASE` that would require being able to drop/create the database. Instead you can use `--query` against some existing data to actually get some results, I think you still need `--create-schema=$MYSQL_DATABASE` for it to connect to the appropriate database. With the tool configured correctly I see results like this:
```
root@7055b351-de49-4c90-a17e-3c241bb3fa7f:/# mysqlslap -h $MYSQL_HOST --user=$MYSQL_USER --password=$MYSQL_PWD --delimiter=";" --create-schema=$MYSQL_DATABASE --create="CREATE TABLE t (a int);INSERT INTO t VALUES (5)" --query="SELECT * FROM t" --iterations=100
Benchmark
       	Average number of seconds to run all queries: 0.000 seconds
       	Minimum number of seconds to run all queries: 0.000 seconds
       	Maximum number of seconds to run all queries: 0.001 seconds
       	Number of clients running queries: 1
       	Average number of queries per client: 1
```

As you can see a simply query takes almost no time to execute, this was over 100 iterations but setting it higher still sees stable query performance.
